### PR TITLE
Ensure files are closed when destructing report objects.

### PR DESCRIPTION
**Pull request by jakobluettgau**
This might provide some more desirable behavior when working with many logs in a single process such as raised by #578.

This pull requests adds hooks for `__del__`, `__enter__`,`__exit__` to call a cleanup routine to ensure opened log files are also closed when the object is destroyed.

### CODE REVIEW

File: darshan-util/pydarshan/darshan/report.py [hunk 1]
```diff
@@ -1058,6 +1058,25 @@ def to_json(self):
...
+        self._cleanup()
 
 
+    def __enter__(self):
+        """ Satisfy API for use with context manager (e.g., with-statement) """
+        return self
 
+    def __exit__(self, type, value, traceback):
+        """ Cleanup when used by context manager (e.g., with-statement) """
+        self._cleanup()
```
Comment (tylerjereddy) [inline darshan-util/pydarshan/darshan/report.py:1082]: We could bump the coverage of this PR up to 100 % with a test for context manger and the issue above at the same time:

```diff
--- a/darshan-util/pydarshan/tests/test_report.py
+++ b/darshan-util/pydarshan/tests/test_report.py
@@ -4,6 +4,7 @@
 """Tests for `pydarshan` package."""
 
 import copy
+import pickle
 
 import pytest
 import numpy as np
@@ -357,3 +358,17 @@ class TestDarshanRecordCollection:
             actual_fct_df = actual_records["fcounters"]
             assert_frame_equal(actual_ct_df, expected_ct_df)
             assert_frame_equal(actual_fct_df, expected_fct_df)
+
+def test_file_closure():
+    # regression test for gh-578
+    with darshan.DarshanReport("examples/example-logs/ior_hdf5_example.darshan") as report:
+        # you cannot serialize a report object
+        # until after you close its associated
+        # file handle; check that __del__ cleans up
+        # poperly
+        report.__del__()
+        pickle.dumps(report)
+        # let's also check that we can properly exit
+        # context when report.log has been removed
+        # (happens with pytest sometimes)
+        del(report.log)

```

**Review by tylerjereddy — COMMENTED** [review]
Makes sense, added a few suggestions

**Comment by jakobluettgau** [comment]
Sounds good, I have pushed updates including the two changes making `_cleanup()` more graceful, as well as adding the test case for use in context managers.

**Comment by tylerjereddy** [comment]
merged, thanks

**Pull request merged**